### ISSUE TITLE
error 401

### ISSUE THREAD

**Issue by 211260fg** [gautamsi/ews-javascript-api#15]
I'm not sure what the issue is, but I get a 401 error when I call GetUserAvailability (I didn't use the autodiscover function). Could this be a problem with authentication on the server side?
Here is the code:

```
var ews = require('ews-javascript-api');

var exch = new ews.ExchangeService(ews.ExchangeVersion.Exchange2010);
exch.Credentials = new ews.ExchangeCredentials("xxxxx", "xxxxxxxx", "webmail.xxxx.be");
exch.Url = new ews.Uri("https://webmail.xxxx.be/Ews/Exchange.asmx");

var attendee =[new ews.AttendeeInfo("xxxxx@xxxx.xxxx")];
var timeWindow = new ews.TimeWindow(ews.DateTime.Now, new ews.DateTime(ews.DateTime.Now.TotalMilliSeconds + ews.TimeSpan.FromHours(48).asMilliseconds()));

exch.GetUserAvailability(attendee, timeWindow, ews.AvailabilityData.FreeBusyAndSuggestions).then(function (availabilityResponse) {
    console.log("succes");
  }, function (errors) {
    console.log(errors);
  });
```
and here is the error:
```
[ServiceObjectPropertyDefinition.ctor] uri is null or empty
[ServiceObjectPropertyDefinition.ctor] uri is null or empty
[ServiceObjectPropertyDefinition.ctor] uri is null or empty
[ServiceObjectPropertyDefinition.ctor] uri is null or empty
sending ews request
{ url: 'https://webmail.xxxx.be/Ews/Exchange.asmx',
  headers: 
   { 'Content-Type': 'text/xml; charset=utf-8',
     Accept: 'text/xml',
     Authorization: 'Basic dmVya2xxxxxxxxxxXXvMQ==' },
  type: 'POST',
  data: '<soap:Envelope xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><soap:Header><t:RequestServerVersion xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types" Version="Exchange2010"></t:RequestServerVersion><t:TimeZoneContext xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:TimeZoneDefinition xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types" Name="UTC" Id="UTC"></t:TimeZoneDefinition></t:TimeZoneContext></soap:Header><soap:Body><m:GetUserAvailabilityRequest xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages"><m:MailboxDataArray xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages"><t:MailboxData xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:Email xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:Address xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">xxxxx@xxxx.xxxx</t:Address></t:Email><t:AttendeeType xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">Required</t:AttendeeType><t:ExcludeConflicts xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">false</t:ExcludeConflicts></t:MailboxData></m:MailboxDataArray><t:FreeBusyViewOptions xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:TimeWindow xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:StartTime xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">2016-03-14T00:00:00</t:StartTime><t:EndTime xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">2016-03-16T00:00:00</t:EndTime></t:TimeWindow><t:MergedFreeBusyIntervalInMinutes xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">30</t:MergedFreeBusyIntervalInMinutes><t:RequestedView xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">Detailed</t:RequestedView></t:FreeBusyViewOptions><t:SuggestionsViewOptions xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:GoodThreshold xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">25</t:GoodThreshold><t:MaximumResultsByDay xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">10</t:MaximumResultsByDay><t:MaximumNonWorkHourResultsByDay xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">0</t:MaximumNonWorkHourResultsByDay><t:MeetingDurationInMinutes xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">60</t:MeetingDurationInMinutes><t:MinimumSuggestionQuality xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">Fair</t:MinimumSuggestionQuality><t:DetailedSuggestionsWindow xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types"><t:StartTime xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">2016-03-14T00:00:00</t:StartTime><t:EndTime xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">2016-03-16T00:00:00</t:EndTime></t:DetailedSuggestionsWindow></t:SuggestionsViewOptions></m:GetUserAvailabilityRequest></soap:Body></soap:Envelope>' }
Error in calling service, error code:401
```

**Comment by gautamsi** [comment]
this means password is incorrect. it may not be in correct format "domain\user" or user@domain.be" whichever works with Exchange server. 

hopefully you have also  masked Authorization header, if not please change password. 

if it is onprem server can you check if webservice virtual directory is configured with basic authentication. If it is NTLM you can use example mentioned in readme. If it is published behind ISA/TMG, you can use cookies authentication. A gist is created for each example.

**Comment by 211260fg** [comment]
I'm not sure what you mean by "correct format", do you mean the string of attendee? (as you can tell, this is the first time I'm working with Exchange, so I'm a bit clueless)

I'll check the authentication later with the IT guys.
Thanks for the quick response anyways!

**Comment by gautamsi** [comment]
try new` ews.ExchangeCredentials("user@domain.be", "password");`
user@domain.be is user's UPN. 
or try `ews.ExchangeCredentials("domain\user", "password");`
domain\user is how you login to your machine in domain environment.

**Comment by gautamsi** [comment]
did it work for you?

**Comment by 211260fg** [comment]
The solution you gave didn't work, but the problem is definitely with the authentication of the server.
I just haven't had time to work it out yet. Still, thanks for the support :)

**Comment by gautamsi** [comment]
you can figure out yourself
- browse https://webmail.xxxx.be/Ews/Exchange.asmx

if it redirects you to a web form with username and password field. it is cookies authentication with TMG (usual configuration)

if it prompts you to enter username and password on a small dialog with username and password field, it should be basic or NTML. basic works out of box with ews-javascript-api. 

for Cookies authentication (form based web page) use: https://gist.github.com/gautamsi/4aec3307942de9087e89

for NTML authentication use - https://gist.github.com/gautamsi/28211eda711e3e7dc04c

reopen this case if you face any issues.

**Comment by nofear87** [comment]
I have the same problem. when I navigate to exchange.asmx I can insert my credentials (look like a htaccess prompt) and come to an site where I got informed that I have created a service. But when using the script I got 401.

'''
Error in calling service, error code:401
server: Microsoft-IIS/8.5
request-id: f2bb00de-6601-445f-be6e-65bd67abde37
www-authenticate: Negotiate, NTLM
x-powered-by: ASP.NET
x-feserver: EXCHANGE02
date: Thu, 22 Sep 2016 12:52:52 GMT
content-length: 0
'''
